src/main/java/com/aws/greengrass/lifecyclemanager/KernelAlternatives.java

```diff
@@ -392,6 +417,98 @@ public void cleanupLaunchDirectoryLinks() {
         cleanupLaunchDirectoryLink(getNewDir());
     }
 
+    /**
+     * Prepare for bootstrapping in the context of a rollback deployment, if such bootstrapping is required.
+     *
+     * @param context Context instance for the rollback configuration
+     * @param deploymentDirectoryManager DeploymentDirectoryManager instance for obtaining path information
+     * @param bootstrapManager BootstrapManager instance for managing pending bootstrap tasks
+     * @return true if bootstrapping is required during rollback, otherwise false
+     */
+    public boolean prepareBootstrapOnRollbackIfNeeded(Context context,
+                                                      DeploymentDirectoryManager deploymentDirectoryManager,
+                                                      BootstrapManager bootstrapManager) {
+        Configuration rollbackConfig = new Configuration(context);
+        try {
+            rollbackConfig.read(deploymentDirectoryManager.getSnapshotFilePath());
+        } catch (IOException exc) {
+            logger.atError().log("Failed to read rollback snapshot config", exc);
+            return false;
+        }
+        boolean bootstrapOnRollbackRequired = false;
+        try {
+            // Check if we need to execute component bootstrap steps during the rollback deployment.
+            final Set<String> componentsToExclude =
+                    getComponentsToExcludeFromBootstrapOnRollback(bootstrapManager, rollbackConfig);
+            bootstrapOnRollbackRequired = bootstrapManager.isBootstrapRequired(rollbackConfig.toPOJO(),
```

Comment:
Inside `isBootstrapRequired`, if a geneic component can be loaded, it will check if the bootstrap has already been executed in the past - https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/4db3bfb5999db9b9bca09b59db475e292ae54610/src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java#L272-L303.

We won't load any service here because no service is running. We just rely on the rollback config.

For components A, B with bootstraps, is it possible if 
- First deployment deploys A v1 successfully;
- Second deployment deploys B v1 successfully;
- Third deployment deploys B v2 which fails and attempts to roll back;
  - rollback config finds A v1 and B v1
  - isBootstrapRequired determines that both A and B bootstraps need to execute on rollback, but only B v2 should be rolling back

Reply:
On testing this scenario, during rollback from the 3rd deployment which fails, I found `isBootstrapRequired` determines that only the bootstrap for B v1 is needed.

For component A, the debug log message is `Bootstrap is not required: bootstrap step unchanged`.  This means that, in the context of the rollback,

1) Nucleus is able to locate the bootstrap config for component A in the failed target deployment
2) Nucleus is able to locate the bootstrap config for component A in the rollback deployment
3) these two bootstrap configs are equal (according to `GenericExternalService.serviceLifecycleBootstrapEquals()`)

This makes sense to me, because the 3rd deployment made no change to the config for component A.